lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -276,536 +344,505 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
+        // If there are more individual shader compile in future, we can add more stages here
+        userDataValue = DescRelocMagic | node.innerTable[0].set;
       }
+      break;
 
-      case ResourceNodeType::DescriptorBuffer:
-      case ResourceNodeType::DescriptorResource:
-      case ResourceNodeType::DescriptorSampler:
-      case ResourceNodeType::DescriptorTexelBuffer:
-      case ResourceNodeType::DescriptorFmask:
-        // Where a descriptor appears in the top-level table, it is accesssed
-        // via the spill table, rather than directly placed in sgprs.
-        assert(needSpill);
-        if (intfData->spillTable.offsetInDwords == InvalidValue)
-          intfData->spillTable.offsetInDwords = node->offsetInDwords;
-        m_pipelineState->getPalMetadata()->setUserDataSpillUsage(node->offsetInDwords);
-        break;
+    case ResourceNodeType::PushConst:
+    case ResourceNodeType::DescriptorBufferCompact:
+      break;
 
-      case ResourceNodeType::PushConst:
-      case ResourceNodeType::DescriptorBufferCompact: {
-        argTys.push_back(VectorType::get(Type::getInt32Ty(*m_context), node->sizeInDwords));
-        m_pipelineState->getPalMetadata()->setUserDataEntry(m_shaderStage, userDataIdx, node->offsetInDwords,
-                                                            node->sizeInDwords);
-        userDataIdx += node->sizeInDwords;
-        break;
-      }
-      default: {
-        llvm_unreachable("Should never be called!");
-        break;
-      }
-      }
-    } else {
-      if (needSpill && intfData->spillTable.offsetInDwords == InvalidValue)
-        intfData->spillTable.offsetInDwords = node->offsetInDwords;
-      m_pipelineState->getPalMetadata()->setUserDataSpillUsage(node->offsetInDwords);
+    default:
+      // Any other descriptor type in the root table is accessed via the spill table.
+      mustSpill = true;
     }
-  }
 
-  // Internal user data
-  if (needSpill && useFixedLayout) {
-    // Add spill table
-    assert(intfData->spillTable.offsetInDwords != InvalidValue);
-    assert(userDataIdx <= (InterfaceData::MaxCsUserDataCount + InterfaceData::CsStartUserData));
-    while (userDataIdx <= (InterfaceData::MaxCsUserDataCount + InterfaceData::CsStartUserData)) {
-      *inRegMask |= 1ull << argTys.size();
-      argTys.push_back(Type::getInt32Ty(*m_context));
-      ++userDataIdx;
+    if (useFixedLayout && userDataIdx != node.offsetInDwords + InterfaceData::CsStartUserData) {
+      // With useFixedLayout, we need a padding arg before the node's arg.
+      assert(node.offsetInDwords + InterfaceData::CsStartUserData > userDataIdx);
+      userDataArgs.push_back(UserDataArg(
+          VectorType::get(builder.getInt32Ty(), node.offsetInDwords + InterfaceData::CsStartUserData - userDataIdx),
+          Util::Abi::UserDataMapping::Invalid, nullptr, /*isPadding=*/true));
+      userDataIdx = node.offsetInDwords + InterfaceData::CsStartUserData;
     }
```

Comment:
This may have to be a loop, incrementing userDataIdx by 1 each time?

Reply:
I think it's fine, because the UserDataArg constructor's second arg is the number of dwords you want to use up.

Reply:
Ah I missed that, thanks.